plutus-core/plutus-core/src/PlutusCore/Builtin/Meaning.hs

```diff
@@ -116,33 +122,52 @@ function and the 'TypeScheme' of the built-in function will be derived automatic
 monomorphic and simply-polymorphic cases no types need to be specified at all.
 -}
 
+type family Length xs where
+    Length '[]       = 'Z
+    Length (_ ': xs) = 'S (Length xs)
+
 type family GetArgs a :: [GHC.Type] where
     GetArgs (a -> b) = a ': GetArgs b
     GetArgs _        = '[]
 
 -- | A class that allows us to derive a monotype for a builtin.
 class KnownMonotype val args res a | args res -> a, a -> res where
     knownMonotype :: TypeScheme val args res
+    knownMonoruntime :: RuntimeScheme (Length args)
+    toImmediateF :: FoldArgs args res -> ToDenotationType val (Length args)
+    toDeferredF :: ReadKnownM (FoldArgs args res) -> ToDenotationType val (Length args)
 
 -- | Once we've run out of term-level arguments, we return a 'TypeSchemeResult'.
 instance (res ~ res', KnownTypeAst (UniOf val) res, MakeKnown val res) =>
             KnownMonotype val '[] res res' where
     knownMonotype = TypeSchemeResult
+    knownMonoruntime = RuntimeSchemeResult
+    toImmediateF = makeKnown (Just ())
+    toDeferredF getRes = liftEither getRes >>= makeKnown (Just ())
+    {-# INLINE toDeferredF #-}
 
 -- | Every term-level argument becomes as 'TypeSchemeArrow'.
 instance
         ( KnownTypeAst (UniOf val) arg, MakeKnown val arg, ReadKnown val arg
         , KnownMonotype val args res a
         ) => KnownMonotype val (arg ': args) res (arg -> a) where
     knownMonotype = TypeSchemeArrow knownMonotype
+    knownMonoruntime = RuntimeSchemeArrow $ knownMonoruntime @val @args @res
+    toImmediateF f = fmap (toImmediateF @val @args @res . f) . readKnown (Just ())
+    toDeferredF getF = \arg ->
+        pure . toDeferredF @val @args @res $! getF <*> readKnown (Just ()) arg
+    {-# INLINE toDeferredF #-}
 
 -- | A class that allows us to derive a polytype for a builtin.
-class KnownPolytype (binds :: [Some TyNameRep]) val args res a | args res -> a, a -> res where
+class KnownMonotype val args res a =>
```

Comment:
I was a little surprised by this! To be clear, this is the monotype that corresponds to the type after stripping off the polymorphic arguments, right?

Reply:
After stripping off the `all` constructors binding type variables. There's a comment about the naming:

> We use two type classes for automatic derivation of type schemes: 'KnownMonotype' and
> 'KnownPolytype'. The terminology is due to https://en.wikipedia.org/wiki/Hindley%E2%80%93Milner_type_system#The_Hindley%E2%80%93Milner_type_system

I should remove the "monomorphic" word from that Note (right down below the quote), though.

Now `KnownMonotype` is a superclass of `KnownPolytype`, because we don't need to have `toImmediateF` and `toDeferredF` in `KnownPolytype`, since for polytypes those are the same as for their corresponding monotypes. Hence I just added the constraint. `KnownPolytype` only allows us to `all`-bind type variables and after than it's just

```haskell
instance KnownMonotype val args res a => KnownPolytype '[] val args res a where
    knownPolytype = knownMonotype
    knownPolyruntime = knownMonoruntime @val @args @res
``` 

so no reason to thread `toImmediateF` and `toDeferredF` through `KnownPolytype` given that none of them is of any relevance to any `all`-bound type variables.